stan/math/prim/mat/fun/gp_matern32_cov.hpp

```diff
@@ -0,0 +1,284 @@
+ *   exp(-\sqrt{3}\sqrt{\sum_{k=1}^{K}\frac{d(x, x')^2}{l_k^2}}) \f]
+ *
+ * where \f$d(x, x')\f$ is the Euclidean distance.
+ *
+ * @tparam T_x type for each scalar
+ * @tparam T_s type of element of parameter sigma
+ * @tparam T_l type of each length scale parameter
+ *
+ * @param x std::vector of Eigen vectors of scalars.
+ * @param length_scale std::vector of length scales
+ * @param sigma standard deviation that can be used in stan::math::square
+ * @throw std::domain error if sigma <= 0, l <= 0, or x is nan or inf
+ */
+template <typename T_x, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern32_cov(const std::vector<Eigen::Matrix<T_x, -1, 1>> &x,
+                const T_s &sigma, const std::vector<T_l> &length_scale) {
+  using std::exp;
+
+  size_t x_size = x.size();
+  Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+  size_t l_size = length_scale.size();
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_matern32_cov", "x", x[n]);
+
+  check_positive_finite("gp_matern32_cov", "magnitude", sigma);
+  check_positive_finite("gp_matern32_cov", "length scale", length_scale);
+
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_matern32_cov", "length scale", length_scale[n]);
+
+  check_size_match("gp_matern32_cov", "x dimension", x[0].size(),
+                   "number of length scales", l_size);
+
+  double sigma_sq = square(sigma);
+  double root_3 = sqrt(3.0);
+  double neg_root_3 = -1.0 * sqrt(3.0);
+
+  std::vector<Eigen::Matrix<typename return_type<T_x, T_l>::type, -1, 1>> x_new
+      = divide_columns(x, length_scale);
+
+  for (size_t i = 0; i < x_size; ++i) {
+    for (size_t j = i; j < x_size; ++j) {
+      typename return_type<T_x, T_s, T_l>::type distance
+          = sqrt(squared_distance(x_new[i], x_new[j]));
+      cov(i, j)
+          = sigma_sq * (1.0 + root_3 * distance) * exp(neg_root_3 * distance);
+      cov(j, i) = cov(i, j);
+    }
+  }
+  return cov;
+}
+
+/**
+ * Returns a Matern 3/2 cross covariance matrix
+ *
+ * \f[ k(x, x') = \sigma^2(1 +
+ *  \frac{\sqrt{3}d(x, x')}{l})exp(-\sqrt{3}\frac{d(x, x')}{l})
+ * \f]
+ *
+ * where \f$d(x, x')\f$ is the Euclidean distance.
+ *
+ * This function is for the cross covariance matrix neededed
+ * to compute the posterior predictive density.
+ *
+ * @tparam T_x1 first type of scalars contained in vector x1
+ * @tparam T_x2 second type of scalars contained in vector x2
+ * @tparam T_s type of parameter sigma, marginal standard deviation
+ * @tparam T_l type of parameter length scale
+ *
+ * @param x1 std::vector of scalars that can be used in squared_distance
+ * @param x2 std::vector of scalars that can be used in squared_distance
+ *    This function assumes all elements of x1 and x2 are the same length.
```

Comment:
Can we clarify this? I think we should assume all elements of x1 has the same size. Same with X2, but not them both. I'm thinking users from the Stan language can make that mistake. 

Can we add the checks from the other PR where we check that the size of the elements of x1 matches X2?

Reply:
from the other PR, we check where the size matches x1 and x2 implicitly when we check to see that the number of of dimensions are same as 'P' (or D) for each x. I agree, and I'll go ahead and add checks for all, so the user doesn't make this mistake

Reply:
Ok, so here I'm encountering issues, and I have reason to believe this is why the assumption was here in the first place. my compiler is not happy trying to get the `size()` of an element when it's a scalar (because T_x can either be a scalar type or vector type for the 1st and 3rd functions... so I have to do some tedious metaprogramming or something)

Reply:
We have a template meta program for that! I don't know what it is off the top of my head, but it's there! Don't rewrite it unless we really don't have it.

Reply:
We have a template meta program for that! I don't know what it is off the top of my head, but it's there! Don't rewrite it unless we really don't have it.